code/modules/vehicles/mecha/mech_bay.dm

```diff
@@ -52,27 +61,86 @@
 	if(in_range(user, src) || isobserver(user))
 		. += span_notice("The status display reads: Recharge power <b>[siunit(recharge_power, "W", 1)]</b>.")
 
-/obj/machinery/mech_bay_recharge_port/process(seconds_per_tick)
+/obj/machinery/mech_bay_recharge_port/power_change()
+	if(!(machine_stat & NOPOWER))
+		begin_processing()
+	return ..()
+
+
+/**
+ * Attempts to "connect" to a machine at the recharge location, prioritizing mechs.
+ *
+ * Does not connect to objects in the /obj/machinery/power tree. Starts processing if
+ * connection is valid.
+ */
+/obj/machinery/mech_bay_recharge_port/proc/try_connect()
 	if(machine_stat & NOPOWER || !recharge_console)
 		return
-	var/obj/vehicle/sealed/mecha/recharging_mech = recharging_mech_ref?.resolve()
-	if(!recharging_mech)
-		recharging_mech = locate(/obj/vehicle/sealed/mecha) in recharging_turf
-		if(recharging_mech)
-			recharging_mech_ref = WEAKREF(recharging_mech)
+	var/obj/recharging_machine = recharging_machine_ref?.resolve()
+	var/obj/item/stock_parts/cell/electron_cache = recharging_powercell_ref?.resolve()
+
+	if(recharging_machine?.loc != recharging_turf)
+		recharging_machine = null
+		recharging_machine_ref = null
+
+	if(electron_cache?.loc != recharging_machine?.loc)
+		electron_cache = null
+		recharging_powercell_ref = null
+
+	if(!recharging_machine)
+		electron_cache = null
+		var/obj/vehicle/sealed/mecha/exosuit = locate(/obj/vehicle/sealed/mecha) in recharging_turf
+		if(exosuit)
+			recharging_machine = exosuit
+			recharging_machine_ref = WEAKREF(recharging_machine)
+			electron_cache = exosuit.cell
+			recharging_powercell_ref = WEAKREF(electron_cache)
 			recharge_console.update_appearance()
-	if(!recharging_mech?.cell)
+		else
+			for(var/obj/machinery/machine in recharging_turf)
+				if(istype(machine, /obj/machinery/power))
+					continue //nice try
+				electron_cache = locate(/obj/item/stock_parts/cell) in machine.contents
+				if(electron_cache)
+					recharging_machine = machine
+					recharging_machine_ref = WEAKREF(recharging_machine)
+					recharging_powercell_ref = WEAKREF(electron_cache)
+					recharge_console.update_appearance()
+					break
+
+	if(recharging_machine && electron_cache)
+		begin_processing()
+
+/obj/machinery/mech_bay_recharge_port/process(seconds_per_tick)
+	if(machine_stat & NOPOWER || !recharge_console)
+		end_processing()
+		return
+
+	var/obj/recharging_machine = recharging_machine_ref?.resolve()
+	var/obj/item/stock_parts/cell/electron_cache = recharging_powercell_ref?.resolve()
+
+	if(!recharging_machine)
+		end_processing()
+		return
+
+	if(recharging_machine.loc !=  recharging_turf)
```

Comment:
```suggestion
	if(!recharging_machine || (recharging_machine.loc != recharging_turf))
```
I think you want to clear all your refs if the recharge machine weakref fails to resolve?